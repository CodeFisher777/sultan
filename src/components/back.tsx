import arrowLeft from './../assets/images/mobilearrowleft.svg';

export const Back: React.FC = () => {
  return (
    <div data-testid="back" className="namem-back">
      <button>
        <img src={arrowLeft} alt="" />
      </button>
      <p>Назад</p>
    </div>
  );
};
