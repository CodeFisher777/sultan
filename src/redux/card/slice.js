import { createSlice } from '@reduxjs/toolkit';
import { fetchCardRedux } from './asyncActions';

const initialState = {
  items: [],
  status: '',
};

const cardSlice = createSlice({
  name: 'card',
  initialState,
  reducers: {
    setItems(state, action) {
      state.items = action.payload;
    },
  },

  extraReducers: (builder) => {
    builder.addCase(fetchCardRedux.pending, (state) => {
      state.status = 'loading';
      state.items = [];
    });
    builder.addCase(fetchCardRedux.fulfilled, (state, action) => {
      state.items = action.payload;
      state.status = 'success';
    });
    builder.addCase(fetchCardRedux.rejected, (state) => {
      state.status = 'error';
      state.items = [];
    });
  },
});

export const selectCartItemById = (id) => (state) => state.cart.items.find((obj) => obj.id === id);
export const selectCard = (state) => state.card;
export const { setItems } = cardSlice.actions;

export default cardSlice.reducer;
