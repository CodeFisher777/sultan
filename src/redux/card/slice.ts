import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import { fetchCardRedux, fetchAdminCardRedux, fetchRemoveCardRedux } from './asyncActions';
import { Card, CardSliceState, Status } from './types';

const initialState: CardSliceState = {
  items: [],
  status: Status.LOADING,
};

const cardSlice = createSlice({
  name: 'card',
  initialState,
  reducers: {
    setItems(state, action: PayloadAction<Card[]>) {
      state.items = action.payload;
    },
  },

  extraReducers: (builder) => {
    //Продукты по категориям
    builder.addCase(fetchCardRedux.pending, (state) => {
      state.status = Status.LOADING;
      state.items = [];
    });
    builder.addCase(fetchCardRedux.fulfilled, (state, action: PayloadAction<Card[]>) => {
      state.items = action.payload;
      state.status = Status.SUCCESS;
    });
    builder.addCase(fetchCardRedux.rejected, (state) => {
      state.status = Status.ERROR;
      state.items = [];
    });
    //все продукты для админа
    builder.addCase(fetchAdminCardRedux.pending, (state) => {
      state.status = Status.LOADING;
      state.items = [];
    });
    builder.addCase(fetchAdminCardRedux.fulfilled, (state, action: PayloadAction<Card[]>) => {
      state.items = action.payload;
      state.status = Status.SUCCESS;
    });
    builder.addCase(fetchAdminCardRedux.rejected, (state) => {
      state.status = Status.ERROR;
      state.items = [];
    });
    //удаление продукта
    //@ts-ignore
    builder.addCase(fetchRemoveCardRedux.pending, (state, action: PayloadAction<string>) => {
      state.items = state.items.filter((obj) => obj.id !== action.payload);
      state.status = Status.LOADING;
    });
  },
});

export const selectCard = (state: RootState) => state.card;
export const { setItems } = cardSlice.actions;

export default cardSlice.reducer;
