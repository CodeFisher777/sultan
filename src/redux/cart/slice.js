import { createSlice } from '@reduxjs/toolkit';
import { calcTotalPrice } from '../../utils/calcTotalPrice';
import { getCartFromLS } from '../../utils/getCartFromLS';

const initialState = {
  items: [],
  totalPrice: 0,
};

const cartSlice = createSlice({
  name: 'cart',
  initialState,
  reducers: {
    addItem(state, action) {
      const findItem = state.items.find((obj) => obj.id === action.payload.id);
      if (findItem) {
        findItem.count++;
      } else {
        state.items.push({
          ...action.payload,
          count: 1,
        });
      }
      state.totalPrice = calcTotalPrice(state.items);
    },

    minusItem(state, action) {
      const findItem = state.items.find((obj) => obj.id === action.payload);
      if (findItem) {
        state.totalPrice -= findItem.price;
        findItem.count > 1
          ? findItem.count--
          : (state.items = state.items.filter((obj) => obj.id !== action.payload));
      }
    },
    removeItem(state, action) {
      const findItem = state.items.find((obj) => obj.id === action.payload);
      if (findItem) {
        state.totalPrice -= findItem.price * findItem.count;
        state.items = state.items.filter((obj) => obj.id !== action.payload);
      }
    },
    clearItems(state) {
      state.items = [];
      state.totalPrice = 0;
    },
  },
});

export const { addItem, removeItem, minusItem, clearItems } = cartSlice.actions;
export default cartSlice.reducer;
