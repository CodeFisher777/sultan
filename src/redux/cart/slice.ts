import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { calcTotalPrice } from '../../utils/calcTotalPrice';
import { getCartFromLS } from '../../utils/getCartFromLS';
import { RootState } from '../store';
import { CartItem, CartSliceState } from './types';

const initialState: CartSliceState = getCartFromLS();

const cartSlice = createSlice({
  name: 'cart',
  initialState,
  reducers: {
    addItem(state, action: PayloadAction<CartItem>) {
      const findItem = state.items.find((obj) => obj.id === action.payload.id);
      if (findItem) {
        findItem.count++;
      } else {
        state.items.push({
          ...action.payload,
          count: 1,
        });
      }
      state.totalPrice = calcTotalPrice(state.items);
    },
    addOneItem(state, action: PayloadAction<CartItem>) {
      const findItem = state.items.find((obj) => obj.id === action.payload.id);

      if (findItem) {
        findItem.count = action.payload.itemCount;
      } else {
        state.items.push({
          ...action.payload,
          count: action.payload.itemCount,
        });
      }
      state.totalPrice = calcTotalPrice(state.items);
    },

    minusItem(state, action: PayloadAction<string>) {
      const findItem = state.items.find((obj) => obj.id === action.payload);
      if (findItem) {
        state.totalPrice -= findItem.price;
        findItem.count > 1
          ? findItem.count--
          : (state.items = state.items.filter((obj) => obj.id !== action.payload));
      }
    },
    removeItem(state, action: PayloadAction<string>) {
      const findItem = state.items.find((obj) => obj.id === action.payload);
      if (findItem) {
        state.totalPrice -= findItem.price * findItem.count;
        state.items = state.items.filter((obj) => obj.id !== action.payload);
      }
    },
    clearItems(state) {
      state.items = [];
      state.totalPrice = 0;
    },
  },
});
export const selectCart = (state: RootState) => state.cart;
export const selectCartItemById = (id: string) => (state: RootState) =>
  state.cart.items.find((obj) => obj.id === id);
export const { addItem, addOneItem, removeItem, minusItem, clearItems } = cartSlice.actions;
export default cartSlice.reducer;
