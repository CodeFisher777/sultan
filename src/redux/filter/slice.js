import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  categoryId: 'Уход за телом',
  currentPage: 1,
  sort: {
    name: 'Название ▼',
    sortProperty: 'title',
  },
};

const filterSlice = createSlice({
  name: 'filters',
  initialState,
  reducers: {
    setCategoryId(state, action) {
      state.categoryId = action.payload;
    },
    setSort(state, action) {
      state.sort = action.payload;
    },
    setCurrentPage(state, action) {
      state.currentPage = action.payload;
    },
  },
});

export const { setCategoryId, setSort, setCurrentPage } = filterSlice.actions;
export default filterSlice.reducer;
