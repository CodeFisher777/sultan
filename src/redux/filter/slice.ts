import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import { FilterSliceState, Sort, SortPropertyEnum } from './types';

const initialState: FilterSliceState = {
  categoryId: 'Уход за телом',
  currentPage: 1,
  sort: {
    name: 'Название ▼',
    sortProperty: SortPropertyEnum.TITLE_DESC,
  },
  minPrice: 0,
  maxPrice: 0,
};

const filterSlice = createSlice({
  name: 'filters',
  initialState,
  reducers: {
    setCategoryId(state, action: PayloadAction<string>) {
      state.categoryId = action.payload;
    },
    setSort(state, action: PayloadAction<Sort>) {
      state.sort = action.payload;
    },
    setCurrentPage(state, action: PayloadAction<number>) {
      state.currentPage = action.payload;
    },
    setMinPrice(state, action: PayloadAction<number>) {
      state.minPrice = action.payload;
    },
    setMaxPrice(state, action: PayloadAction<number>) {
      state.maxPrice = action.payload;
    },
  },
});
export const selectFilter = (state: RootState) => state.filter;
export const selectCategory = (state: RootState) => state.filter.categoryId;
export const selectSort = (state: RootState) => state.filter.sort;

export const { setCategoryId, setSort, setCurrentPage, setMaxPrice, setMinPrice } =
  filterSlice.actions;
export default filterSlice.reducer;
