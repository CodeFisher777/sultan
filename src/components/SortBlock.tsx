import React from 'react';

import { useDispatch } from 'react-redux';
import { useMediaQuery } from 'react-responsive';
import { selectSort, setSort } from '../redux/filter/slice';
import { Sort, SortPropertyEnum } from '../redux/filter/types';

type SortItem = {
  name: string;
  sortProperty: SortPropertyEnum;
};
type PopupClick = MouseEvent & {
  path: Node[];
};
type SortPopupProps = {
  value: Sort;
};

export const sortList: SortItem[] = [
  { name: 'Название ▼', sortProperty: SortPropertyEnum.TITLE_DESC },
  { name: 'Название ▲', sortProperty: SortPropertyEnum.TITLE_ASC },
  { name: 'Цена ▼', sortProperty: SortPropertyEnum.PRICE_DESC },
  { name: 'Цена ▲', sortProperty: SortPropertyEnum.PRICE_ASC },
];

export const SortBlock: React.FC<SortPopupProps> = React.memo(({ value }) => {
  const dispatch = useDispatch();

  const [open, setOpen] = React.useState(false);
  const sortRef = React.useRef<HTMLDivElement>(null);
  const isMobile = useMediaQuery({ query: '(max-width:950px)' });
  const onClickListItem = (obj: SortItem) => {
    dispatch(setSort(obj));
    setOpen(false);
  };

  React.useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      const _event = event as PopupClick;
      let path = sortRef.current && event.composedPath().includes(sortRef.current);
      if (!path) {
        setOpen(false);
      }
    };
    document.body.addEventListener('click', handleClickOutside);

    return () => {
      document.body.removeEventListener('click', handleClickOutside);
    };
  }, []);
  return (
    <section className="nameandsort">
      <div ref={sortRef} className="sort">
        <div className="sort-label">
          <p>Сортировка:</p>
          <span data-testid="toggle-btn" onClick={() => setOpen(!open)}>
            {value?.name}
          </span>
        </div>
        {open && (
          <div className="sort-popup" data-testid="toggle-elem">
            <ul>
              {sortList.map((obj, i) => (
                <li
                  key={i}
                  onClick={() => onClickListItem(obj)}
                  className={value?.sortProperty === obj.sortProperty ? 'active' : ''}
                >
                  {obj.name}
                </li>
              ))}
            </ul>
          </div>
        )}
      </div>
    </section>
  );
});
