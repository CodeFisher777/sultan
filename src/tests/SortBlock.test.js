import { render, screen, fireEvent } from '@testing-library/react';
import * as actions from '../redux/filter/slice';
import * as reduxHooks from 'react-redux';

import { SortBlock } from '../components/SortBlock';

jest.mock('react-redux');

const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');

describe('SortBlock', () => {
  it('should open popup', () => {
    const component = render(<SortBlock name="Название ▼" sortProperty="title" />);

    const btn = screen.getByTestId('toggle-btn');
    expect(screen.queryByTestId('toggle-elem')).toBeNull();
    fireEvent.click(btn);
    expect(screen.queryByTestId('toggle-elem')).toBeInTheDocument();
    fireEvent.click(btn);
    expect(screen.queryByTestId('toggle-elem')).toBeNull();
  });
});
