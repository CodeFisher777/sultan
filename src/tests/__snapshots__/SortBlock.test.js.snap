// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItemBlock should create CartItemBlock 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="line"
      />
      <div
        class="root"
      >
        <img
          alt=""
          class="root__imgmain"
          src="link"
        />
        <div
          class="root__text"
        >
          <div
            class="root__text__v"
          >
            <img
              alt=""
              src="./images/littlebootle.svg"
            />
            <p>
              45
               мл
            </p>
          </div>
          <p
            class="root__text__name"
          >
            Мыло
          </p>
          <p
            class="root__text__desciption"
          >
            description
          </p>
        </div>
        <span
          class="root__vertline"
        />
        <div
          class="root__block"
        >
          <div
            class="root__buttons"
          >
            <button>
              -
            </button>
            <p>
              2
            </p>
            <button>
              +
            </button>
          </div>
          <span
            class="root__vertline"
          />
          <p
            class="root__price"
          >
            1100
             ₸
          </p>
          <span
            class="root__vertline"
          />
          <button
            class="root__del"
          >
            <img
              alt=""
              src="/images/trash.svg"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="line"
    />
    <div
      class="root"
    >
      <img
        alt=""
        class="root__imgmain"
        src="link"
      />
      <div
        class="root__text"
      >
        <div
          class="root__text__v"
        >
          <img
            alt=""
            src="./images/littlebootle.svg"
          />
          <p>
            45
             мл
          </p>
        </div>
        <p
          class="root__text__name"
        >
          Мыло
        </p>
        <p
          class="root__text__desciption"
        >
          description
        </p>
      </div>
      <span
        class="root__vertline"
      />
      <div
        class="root__block"
      >
        <div
          class="root__buttons"
        >
          <button>
            -
          </button>
          <p>
            2
          </p>
          <button>
            +
          </button>
        </div>
        <span
          class="root__vertline"
        />
        <p
          class="root__price"
        >
          1100
           ₸
        </p>
        <span
          class="root__vertline"
        />
        <button
          class="root__del"
        >
          <img
            alt=""
            src="/images/trash.svg"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SortBlock should open popup 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="nameandsort"
      >
        <div
          class="sort"
          data-testid="toggle-elem"
        >
          <div
            class="sort-label"
          >
            <p>
              Сортировка:
            </p>
            <span
              data-testid="toggle-btn"
            />
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="nameandsort"
    >
      <div
        class="sort"
        data-testid="toggle-elem"
      >
        <div
          class="sort-label"
        >
          <p>
            Сортировка:
          </p>
          <span
            data-testid="toggle-btn"
          />
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
